### PR TITLE
SLLS-12 Enable server notifications

### PR DESCRIPTION
**Pull request by jblievremont**

### CODE REVIEW

**Comment by jblievremont** [comment]
@henryju as discussed at standup, could you please have a look that the current state? This is not the final state - a lot of coverage is missing, and the "last notif timestamp" is not persisted yet.

File: src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      final String label = isSonarCloud ? "SonarCloud" : "SonarQube";
+      ShowMessageRequestParams params = new ShowMessageRequestParams();
+      params.setType(MessageType.Info);
+      params.setMessage(String.format("%s Notification: %s", label, serverNotification.message()));
+      MessageActionItem browseAction = new MessageActionItem();
+      browseAction.setTitle("Open in " + label);
+      params.setActions(Collections.singletonList(browseAction));
+      client.showMessageRequest(params).thenAccept(action -> {
+        if(browseAction.equals(action)) {
+          client.browseTo(serverNotification.link());
```
Comment (henryju) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java:156]: Here you should probably call the telemetry to report that the link has been clicked.

File: src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java [hunk 2]
```diff
@@ -142,7 +150,12 @@ static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) thr
       cancelToken.checkCanceled();
       this.traceLevel = parseTraceLevel(params.getTrace());
 
-      progressManager.setWorkDoneProgressSupportedByClient(ofNullable(params.getCapabilities().getWindow().getWorkDoneProgress()).orElse(false));
+      progressManager.setWorkDoneProgressSupportedByClient(
+        ofNullable(params.getCapabilities())
```
Comment (jblievremont) [inline src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java:154]: Got a NPE when testing in VSCode, I believe this code is a bit defensive but at least it avoids an early crash :smile:

File: src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java [hunk 3]
```diff
@@ -158,15 +171,15 @@ static SonarLintLanguageServer bySocket(int port, Collection<URL> analyzers) thr
       // until https://github.com/microsoft/vscode-languageserver-node/pull/697 is released
       // params.getClientInfo().getName()
       String appName = (String) options.get("appName");
-      String ideVersion = appName + " " + params.getClientInfo().getVersion();
+      String ideVersion = appName + " " + ofNullable(params.getClientInfo()).map(ClientInfo::getVersion).orElse("");
```
Comment (jblievremont) [inline src/main/java/org/sonarsource/sonarlint/ls/SonarLintLanguageServer.java:174]: Weirdly, I also got a NPE here.

File: src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    }
+  }
+
+  private void registerConfigurationIfNeeded(String connectionId, String projectKey) {
+    if (!alreadyHasConfiguration(connectionId, projectKey)) {
+      if(!connections.containsKey(connectionId) || connections.get(connectionId).isDevNotificationsDisabled()) {
+        // Connection is unknown, or has notifications disabled - do nothing
+        return;
+      }
+      logDebugMessage(String.format("Enabling notifications for %s on %s", projectKey, connectionId));
```
Comment (henryju) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java:112]:
```suggestion
      logDebugMessage(String.format("Enabling notifications for project '%s' on connection '%s'", projectKey, connectionId));
```

File: src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+    if (newValue.hasBinding()) {
+      registerConfigurationIfNeeded(newValue.getConnectionId(), newValue.getProjectKey());
+    }
+  }
+
+  private void unregisterConfigurationIfExists(@Nullable String oldConnectionId, @Nullable String oldProjectKey) {
+    Map<String, NotificationConfiguration> configsForOldConnectionId = configurationsByProjectKeyByConnectionId.get(oldConnectionId);
+    if (configsForOldConnectionId != null && configsForOldConnectionId.containsKey(oldProjectKey)) {
+      logDebugMessage(String.format("De-registering notifications for %s on %s", oldProjectKey, oldConnectionId));
```
Comment (henryju) [inline src/main/java/org/sonarsource/sonarlint/ls/connected/notifications/ServerNotifications.java:100]:
```suggestion
      logDebugMessage(String.format("De-registering notifications for project '%s' on connection '%s'", oldProjectKey, oldConnectionId));
```

**Review by henryju — COMMENTED** [review]
Minor suggestion to make logs more readable

**Pull request merged**